### PR TITLE
[MRG] Replace get_bids_fname func for fpath in BIDSPath and root for prefix

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #492 
Closes: #454 

- Replaces `prefix` for `root`
- Replace `get_bids_fname` for `fpath`

BIDSPath will allow "inference" of the full file path in the following situations:

1. `modality is None` and `suffix` is one of 'meg', 'eeg', 'ieeg', 'T1w' -> set `modality` automatically
2. `extension is None`, and `modality` is set -> then try to find the correct file path

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

### CODE REVIEW

File: mne_bids/path.py
Comment (agramfort) [inline mne_bids/path.py:99]: bids_basename -> bids_path

File: mne_bids/path.py
Comment (agramfort) [inline mne_bids/path.py:243]: why do we still have kind?

File: mne_bids/path.py
Comment (agramfort) [inline mne_bids/path.py:107]: new_basename -> new_bids_path

**Comment by hoechenberger** [comment]
I'd like to vote for `root` instead of `bids_root` – we're in BIDS land anyway, why be redundant?

**Comment by agramfort** [comment]
I've been thinking about this too. +0.5 on my side

File: mne_bids/path.py
Comment (adam2392) [reply mne_bids/path.py]: So here, I propose altering internal code that does things like `bids_basename = BIDSPath(...) -> `bids_path = BIDSPath(...)` in a later PR because there will be a huge diff otherwise.

File: doc/whats_new.rst
Comment (hoechenberger) [inline doc/whats_new.rst:80]: So I went back to the latest release (0.4) and BIDSPath didn't exist back then if I'm not mistaken … so no need to note **changes** in BIDSPath here, then
```suggestion
```

File: mne_bids/read.py
Comment (hoechenberger) [inline mne_bids/read.py:336]: Should we change the signature of `read_raw_bids` to use `root` instead of `bids_root` too in this PR? Or is this reserved for a followup? Currently this seems a little inconsistent.

File: mne_bids/read.py
Comment (hoechenberger) [inline mne_bids/read.py:339]: We wanted to stop doing that, no? Or is it still useful? I'm undecided!

Comment (hoechenberger) [reply mne_bids/read.py:339]: Ok I had a look at the code, so we're trying to infer `modality` from the suffix for MEG, EEG, iEEG. Which is … probably fine? But it will lead to an inconsistency when using an anatomy suffix such as `T1w`, right? Shouldn't we add those there, too?

File: doc/whats_new.rst
Comment (agramfort) [reply doc/whats_new.rst:80]: also the public import is :class:`mne_bids.path.BIDSPath` and not :class:`mne_bids.BIDSPath`

**Comment by hoechenberger** [comment]
So I was just trying to work with this, and I found it a bit awkward to get my hands on the directory a subject's data are supposed to be placed in:

```python
from mne_bids import BIDSPath
import pathlib
p = BIDSPath(subject='foo', session='bar', modality='eeg', root='/bids/root')

subject_dir = pathlib.Path(p.fpath).parent
print(subject_dir)
```

- I believe `BIDSPath.fpath` should be a `Path`
- maybe we should have `BIDSPath.parent`

**Comment by agramfort** [comment]
+1 to get pathlib.Path from fpath attribute

>

**Comment by hoechenberger** [comment]
> +1 to get pathlib.Path from fpath attribute
> 

Will add this right after finishing my Crêpe 🥞

File: mne_bids/read.py
Comment (hoechenberger) [reply mne_bids/read.py:336]: No actually, we could / should drop `bids_root` here entirely in favor of `BIDSPath.root`

**Review by hoechenberger — APPROVED** [review]
LGTM, the comments I left should probably be addressed in a future PR.

@agramfort Feel free to merge once CI goes green.

**Comment by hoechenberger** [comment]
Merging on @agramfort's behalf, as he's busy grabbing lunch :D

**Comment by hoechenberger** [comment]
Thanks @adam2392!!

File: mne_bids/path.py
Comment (jasmainak) [inline mne_bids/path.py:451]: sorry I am getting to these late ....

raise an error if it's not provided? why assume silently

File: mne_bids/report.py
Comment (jasmainak) [inline mne_bids/report.py:427]: this is awkward ... so `bids_basename` is now an instance of `BIDSPath`? To avoid confusion and be explicit that it's a path object and not str, I would probably do `bids_basename_path`

Comment (jasmainak) [reply mne_bids/report.py:427]: or do we always mean a path object from now on? And it's an `str` only when we do `bids_path.basename` ?

Comment (adam2392) [reply mne_bids/report.py:427]: Yeah I think the next PR, which I can do today is cosmetic, where we rename everything from `bids_basename` to `bids_path`, and if you want the basename str, then you do `bids_path.basename`.

What do you think?

Comment (jasmainak) [reply mne_bids/report.py:427]: okay, and just to clarify `bids_path` is always a full path for a particular file? or can it also represent a "search path" with unknown entities?

Comment (adam2392) [reply mne_bids/report.py:427]: It's always represented by root and the basename, which then builds up the full path, but not necessarily for a particular file. For example:

```Python
bids_path = BIDSPath(subject='01', session='02', run='01')
print(bids_path)
>>> /sub-01/ses-02/sub-01_ses-02_run-01

bids_path.update(root='/bids_root')
>>> /bids_root/sub-01/ses-02/sub-01_ses-02_run-01

bids_path.update(suffix='meg')
>>> /sub-01/ses-02/meg/sub-01_ses-02_run-01_meg

bids_path.update(extension='.fif')
>>> /sub-01/ses-02/meg/sub-01_ses-02_run-01_meg.fif
```

The search path will use `_find_matching_sidecar` which relies on the entities stored in BIDSPath and a suffix/extension that overrides what's in BIDSPath.

Comment (hoechenberger) [reply mne_bids/report.py:427]: > Yeah I think the next PR, which I can do today is cosmetic, where we rename everything from `bids_basename` to `bids_path`, and if you want the basename str, then you do `bids_path.basename`.

+10

Comment (adam2392) [reply mne_bids/report.py:427]: There should be some more "sensible" error checks within ` BIDSPath._check()`, but perhaps we add them in as they come up?

For example if:

```Python
bids_path.update(suffix='meg')
>>> /sub-01/ses-02/meg/sub-01_ses-02_run-01_meg

# this should produce an error because it's not allowed
bids_path.update(modality='eeg')
>>> /sub-01/ses-02/eeg/sub-01_ses-02_run-01_meg
```

Comment (jasmainak) [reply mne_bids/report.py:427]: is there a public function that takes an incomplete `bids_path` and returns a list of filenames matching that?

Comment (adam2392) [reply mne_bids/report.py:427]: No, but... we should! Can you make an issue describing that flow, and we can hash it out there?

Comment (jasmainak) [reply mne_bids/report.py:427]: shouldn't [this](https://mne.tools/mne-bids/dev/generated/mne_bids.path.get_matched_basenames.html#mne_bids.path.get_matched_basenames) take a `bids_path` ?

**Comment by jasmainak** [comment]
okay I'm all caught up. No more emails please for next 2 days ...

File: mne_bids/path.py
Comment (hoechenberger) [reply mne_bids/path.py:451]: > sorry I am getting to these late ....

Sorry about that, should we hold off with the merging a little longer in the future?

File: mne_bids/report.py
Comment (hoechenberger) [reply mne_bids/report.py:427]: That was my intention originally, but @agramfort suggested we take all entities separately to be closer to how a CLI would work. But good thing you're raising this again, maybe we should reconsider this decision? I'd be +0.75 to allow BIDSPath in `mne_bids.path.get_matched_basenames`

**Comment by agramfort** [comment]
yes I agree that mne_bids.path.get_matched_basenames could be made a
method of BIDSPath

BIDSPath.match()

would return a list of BIDSPath will all fields given.

It basically mean that working with mne-bids means understanding how
to work with BIDSPath instance

**Comment by hoechenberger** [comment]
> yes I agree that mne_bids.path.get_matched_basenames could be made a method of BIDSPath

I didn't even think of that, but that sounds super elegant. We could keep the existing function (maybe make it private?) and add the method to BIDSPath, which invokes the function.

**Comment by agramfort** [comment]
yes make get_matched_basenames private

>

File: mne_bids/path.py
Comment (jasmainak) [reply mne_bids/path.py:451]: go with the flow as long as code has been reviewed by one or two maintainers :) I'll catch up when I can.

**Comment by jasmainak** [comment]
sounds like a good plan. Agree with @agramfort 's idea

**Pull request merged**
